src/main/java/com/atlauncher/utils/CommandExecutor.java

```diff
@@ -0,0 +1,160 @@
+        }
+    }
+
+    private static void printStreamToConsole(InputStream stream) {
+        try {
+            InputStreamReader inputStreamReader = new InputStreamReader(stream);
+            BufferedReader reader = new BufferedReader(inputStreamReader);
+            String line;
+
+            while ((line = reader.readLine()) != null) {
+                    LogManager.info(line);
+            }
+        }
+        catch(Exception e)
+        {
+            LogManager.logStackTrace(e);
+            //throw new RuntimeException(e);
+        }
+    }
+
+    //print the whole thing as 1 message rather than line by line
+    private static void printErrorStreamToConsole(InputStream stream) {
+        try {
+            boolean hasGotFirstContentLine = false;
+            InputStreamReader inputStreamReader = new InputStreamReader(stream);
+            BufferedReader reader = new BufferedReader(inputStreamReader);
+            StringBuilder message = new StringBuilder();
+            String line;
+
+            while ((line = reader.readLine()) != null) {
+                //cut off any initial blank lines
+                if(!line.isEmpty() || hasGotFirstContentLine) {
+                    hasGotFirstContentLine = true;
+                    message.append(line).append(System.lineSeparator());
+                }
+            }
+
+            LogManager.error(message.toString());
+        }
+        catch(Exception e)
+        {
+            LogManager.logStackTrace(e);
+            //throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Substitutes all tokens beginning with '$' (and in all capitals) with the matching key in {@code tokens} if it exists
+     *
+     * @param tokens  All the keys that will be replaced with their value in {@code command}
+     * @param command The string to substitute into
+     * @return The string with all tokens substituted
+     */
+    private static String replaceArgumentTokensForCommand(Map<String, String> tokens, String command) {
+        final Pattern tokenPattern = Pattern.compile("\\$([A-Z_]+)");
+        final StringBuffer result = new StringBuffer();
+        Matcher match = tokenPattern.matcher(command);
+
+        while (match.find()) {
+            final String key = match.group(1);
+
+            if (tokens.containsKey(key)) {
+                match.appendReplacement(result, tokens.get(key).replace("\\", "\\\\"));
+            }
+        }
+
+        match.appendTail(result);
+
+        return result.toString();
+    }
+
+    private static Map<String, String> getCommandArgumentTokensForInstance(Instance instance) {
+        final Map<String, String> result = new HashMap<>();
+        result.put("INST_NAME", instance.getName());
+        result.put("INST_ID", instance.getRootDirectory().getName());
+        result.put("INST_DIR", instance.getRootDirectory().getAbsolutePath());
+        //not a thing in atlauncher
+        //result.put("INST_MC_DIR", "");
```

Comment:
Might be good to add this in even if it's the same as the above `INST_DIR` to keep compatability with MultiMC launch scripts. Then maybe just note it in the settings tab that it's the same